### PR TITLE
Improved RestfulService - 3 branch

### PR DESCRIPTION
**Pull request by dhensby**
See #2157 and #2230

### CODE REVIEW

**Comment by dhensby** [comment]
Although I think this should still get into core, I think we should aim to either remove RestfulService or make it a wrapper from `Guzzle` as there are a lot of parts of this class I'd like to re-write.

I know @chillu was in favour of taking in Guzzle instead.

In fact, I think we should probably pull `RestfulService` out of core. It's only used ONCE in the entire framework (and CMS) and that's for requesting OEMBED responses.

It's a lot of code (and tests) to maintain for one function

**Comment by wilr** [comment]
+1 for using Guzzle rather than maintaining our own. @dhensby something for master I guess so happy for this to be included for 3.2? 4.0 I'd imagine deprecating the class is acceptable.

**Comment by dhensby** [comment]
Ok, rebased to make this mergable

File: api/RestfulService.php [hunk 1]
```diff
@@ -164,25 +243,38 @@ protected function constructURL(){
...
+		else {
+			$cacheExpire = 0;
+		}
 
 		// Check for unexpired cached feed (unless flush is set)
 		//assume any cache_expire that is 0 or less means that we dont want to
 		// cache
-		if($this->cache_expire > 0 && !self::$flush
+		if ($canCache && $this->cache_expire > 0 && !self::$flush
 				&& @file_exists($cache_path)
```
Comment (tractorcow) [inline api/RestfulService.php]: Why don't we use the SS_Cache api here?

**Comment by tractorcow** [comment]
I've had my eye on guzzle replacement :) One day someone will have time I hope.

Even better would be to move this out of core into a module.

File: api/RestfulService.php (hunk 1 above)
Comment (dhensby) [inline api/RestfulService.php]: This point was made previously and some of the work has been done here: https://github.com/silverstripe/silverstripe-framework/pull/3471

It wasn't something I was focussing on when I initially made these improvements and this is just a rebase of old work.

My view has shifted to getting people to use guzzle over this and adding a RestfulService poly-fill for those that want legacy support.

Core only uses RestfulService once in Oembed.php and that could be done with a raw curl request instead (as it's so trivial).

I'd therefore opt to remove any kind of client from core and users can require guzzle as needed

**Comment by stevie-mayhew** [comment]
Have you seen https://github.com/colymba/silverstripe-restfulapi - we've been using it in favour of the default RestfulService as it seems to work a bit better :smile:

**Comment by dhensby** [comment]
@stevie-mayhew do you mean instead of `RestfulServer`?

**Comment by stevie-mayhew** [comment]
Bah. Sorry. Continue.

**Comment by tractorcow** [comment]
Is it ok if we address this for 3.3? Is this PR semver?

**Comment by dhensby** [comment]
Yes, this is semver IMO

File: api/RestfulService.php [hunk 2]
```diff
@@ -139,9 +139,62 @@ public function basicAuth($username, $password) {
 
 	/**
 	 * Set a custom HTTP header
+	 * @deprecated
 	 */
 	public function httpHeader($header) {
-		$this->customHeaders[] = $header;
+		Deprecation::notice('3.2', 'Use "addHeader" instead');
```
Comment (sminnee) [inline api/RestfulService.php]: Should probably be 3.4 now...

Comment (sminnee) [inline api/RestfulService.php]: Or 4.0? I can't remember how we decided to mark that...

**Comment by sminnee** [comment]
@dhensby did you want to incorporate these changes or move to deprecated `RestfulService` entirely in SS4? I think we should do 1 or the other...

File: api/RestfulService.php (hunk 2 above)
Comment (tractorcow) [inline api/RestfulService.php]: 4.0, as that is the version this deprecated api will be removed in.

**Comment by tractorcow** [comment]
+1 for deprecating. We should add the deprecation notice in 3 branch, and physically remove it in 4.0.

**Comment by dhensby** [comment]
deprecate

**Pull request closed without merging**